File: Usuario.py
"""
Clase Usuario - Representa un usuario que envía trabajos a la impresora
"""
import multiprocessing
import time


class Usuario:
    """
    Representa un usuario del sistema de impresión
    """
    
    def __init__(self, id_usuario, num_trabajos):
        """
        Constructor del usuario
        
        Args:
            id_usuario: Identificador único del usuario
            num_trabajos: Número de trabajos que enviará este usuario
        """
        self.id_usuario = id_usuario
        self.num_trabajos = num_trabajos
    
    def enviar_trabajos(self, semaforo):
        """
        Envía los trabajos a la impresora de forma secuencial
        Cada trabajo requiere adquirir el semáforo antes de imprimir
        
        Args:
            semaforo: Semáforo que controla el acceso a la impresora
        """
        for num_trabajo in range(1, self.num_trabajos + 1):
            print(f"[Usuario-{self.id_usuario}] Solicita la impresora...")
            
            # Adquirir el semáforo (esperar si está ocupado)
            semaforo.acquire()
            
            try:
                # Sección crítica: imprimir trabajo
                print(f"[Usuario-{self.id_usuario}] Imprimiendo trabajo {num_trabajo}")
                time.sleep(2)  # Simula el tiempo de impresión
                print(f"[Usuario-{self.id_usuario}] Termina trabajo {num_trabajo}")
            finally:
                # Liberar el semáforo
                semaforo.release()
            
            # Pequeña pausa entre trabajos del mismo usuario
            time.sleep(0.5)
    
    def __str__(self):
        """
        Representación en string del usuario
        """
        return f"Usuario-{self.id_usuario} ({self.num_trabajos} trabajos)"


def proceso_usuario(id_usuario, num_trabajos, semaforo):
    """
    Función que se ejecuta como proceso independiente
    Crea una instancia de Usuario y ejecuta sus trabajos
    
    Args:
        id_usuario: ID del usuario
        num_trabajos: Número de trabajos a enviar
        semaforo: Semáforo compartido para sincronización
    """
    usuario = Usuario(id_usuario, num_trabajos)
    usuario.enviar_trabajos(semaforo)

File: Usuario.java
package javaVersion;

import java.io.*;

/**
 * Clase Usuario - Proceso independiente que simula un trabajador
 * enviando trabajos a la impresora compartida
 */
public class Usuario {
    
    public static void main(String[] args) {
        // Validar argumentos
        if (args.length < 4) {
            System.err.println("Error: Faltan argumentos");
            System.err.println("Uso: java Usuario <id> <numTrabajos> <archivoSemaforo> <archivoCola>");
            System.exit(1);
        }
        
        // Parsear argumentos
        int idUsuario = Integer.parseInt(args[0]);
        int numTrabajos = Integer.parseInt(args[1]);
        String archivoSemaforo = args[2];
        String archivoCola = args[3];
        
        // Cada usuario procesa sus trabajos
        for (int trabajo = 1; trabajo <= numTrabajos; trabajo++) {
            try {
                // Solicitar acceso a la impresora
                System.out.println("[Usuario-" + idUsuario + "] Solicita la impresora...");
                
                // ADQUIRIR SEMÁFORO (esperar hasta conseguir el turno)
                adquirirSemaforo(archivoSemaforo);
                
                // SECCIÓN CRÍTICA: Solo un proceso puede estar aquí
                System.out.println("[Usuario-" + idUsuario + "] Imprimiendo trabajo " + trabajo);
                
                // Enviar trabajo a la cola de impresión
                String descripcionTrabajo = "Usuario-" + idUsuario + " Trabajo-" + trabajo;
                enviarTrabajoACola(archivoCola, descripcionTrabajo);
                
                // Simular tiempo de impresión
                Thread.sleep(2000);
                
                System.out.println("[Usuario-" + idUsuario + "] Termina trabajo " + trabajo);
                
                // LIBERAR SEMÁFORO (devolver el turno)
                liberarSemaforo(archivoSemaforo);
                
                // Pausa entre trabajos del mismo usuario
                Thread.sleep(500);
                
            } catch (IOException e) {
                System.err.println("[Usuario-" + idUsuario + "] Error de E/S: " + e.getMessage());
            } catch (InterruptedException e) {
                System.err.println("[Usuario-" + idUsuario + "] Proceso interrumpido");
            }
        }
    }
    
    /**
     * Adquiere el semáforo (espera activa hasta conseguirlo)
     * @param archivo Nombre del archivo que actúa como semáforo
     */
    private static void adquirirSemaforo(String archivo) throws IOException, InterruptedException {
        File semaforo = new File(archivo);
        
        // Intentar crear el archivo (operación atómica)
        // Si ya existe, esperar y reintentar (polling)
        while (!semaforo.createNewFile()) {
            Thread.sleep(100); // Espera 100ms antes de reintentar
        }
        // Si llegamos aquí, hemos conseguido crear el archivo = tenemos el semáforo
    }
    
    /**
     * Libera el semáforo borrando el archivo
     * @param archivo Nombre del archivo que actúa como semáforo
     */
    private static void liberarSemaforo(String archivo) {
        File semaforo = new File(archivo);
        semaforo.delete();
    }
    
    /**
     * Envía un trabajo a la cola de impresión
     * @param archivoCola Archivo que contiene la cola de trabajos
     * @param trabajo Descripción del trabajo a enviar
     */
    private static void enviarTrabajoACola(String archivoCola, String trabajo) throws IOException {
        try (FileWriter fw = new FileWriter(archivoCola, true);
             BufferedWriter bw = new BufferedWriter(fw)) {
            bw.write(trabajo + "\n");
        }
    }
}